numpy/f2py/cfuncs.py

```diff
@@ -621,71 +630,127 @@
-        { STRINGCOPYN(PyArray_DATA(arr),str,PyArray_NBYTES(arr)); }
-    return 1;
+    if (PyArray_Check(obj) && (!((arr = (PyArrayObject *)obj) == NULL))) {
+        string buf = PyArray_DATA(arr);
+        npy_intp n = len;
+        if (n == -1) {
+            /* Assuming null-terminated str. */
+            n = strlen(str);
+        }
+        if (n > PyArray_NBYTES(arr)) {
+            n = PyArray_NBYTES(arr);
+            if (n == 1 && buf[0] == '\0') {
+                /* handle the numpy.array(b'') case */
+                n = 0;
+            }
+        }
+        STRINGCOPYN(buf, str, n);
+        return 1;
+    }
 capi_fail:
     PRINTPYOBJERR(obj);
-    PyErr_SetString(#modulename#_error,\"try_pyarr_from_string failed\");
+    PyErr_SetString(#modulename#_error, \"try_pyarr_from_string failed\");
     return 0;
 }
 """
 needs['string_from_pyobj'] = ['string', 'STRINGMALLOC', 'STRINGCOPYN']
 cfuncs['string_from_pyobj'] = """\
+/*
+  Create a new string buffer from a Python string-like object.
+
+  The string buffer has given size (len) or the size of inistr when len==-1.
+
+  The string buffer may or may not be null-terminated.
+ */
 static int
-string_from_pyobj(string *str,int *len,const string inistr,PyObject *obj,const char *errmess)
+string_from_pyobj(string *str, int *len, const string inistr, PyObject *obj,
+                  const char *errmess)
 {
     PyArrayObject *arr = NULL;
     PyObject *tmp = NULL;
+    string buf = NULL;
+    npy_intp n = -1;
 #ifdef DEBUGCFUNCS
-fprintf(stderr,\"string_from_pyobj(str='%s',len=%d,inistr='%s',obj=%p)\\n\",(char*)str,*len,(char *)inistr,obj);
+fprintf(stderr,\"string_from_pyobj(str='%s',len=%d,inistr='%s',obj=%p)\\n\",
+               (char*)str, *len, (char *)inistr, obj);
 #endif
     if (obj == Py_None) {
-        if (*len == -1)
-            *len = strlen(inistr); /* Will this cause problems? */
-        STRINGMALLOC(*str,*len);
-        STRINGCOPYN(*str,inistr,*len+1);
-        return 1;
+        n = strlen(inistr);
+        buf = inistr;
     }
-    if (PyArray_Check(obj)) {
+    else if (PyArray_Check(obj)) {
         if ((arr = (PyArrayObject *)obj) == NULL)
             goto capi_fail;
         if (!ISCONTIGUOUS(arr)) {
-            PyErr_SetString(PyExc_ValueError,\"array object is non-contiguous.\");
+            PyErr_SetString(PyExc_ValueError,
+                            \"array object is non-contiguous.\");
             goto capi_fail;
         }
-        if (*len == -1)
-            *len = (PyArray_ITEMSIZE(arr))*PyArray_SIZE(arr);
-        STRINGMALLOC(*str,*len);
-        STRINGCOPYN(*str,PyArray_DATA(arr),*len+1);
-        return 1;
-    }
-    if (PyBytes_Check(obj)) {
-        tmp = obj;
-        Py_INCREF(tmp);
-    }
-    else if (PyUnicode_Check(obj)) {
-        tmp = PyUnicode_AsASCIIString(obj);
+        n = PyArray_NBYTES(arr);
+        buf = PyArray_DATA(arr);
+        if (n == 1 && buf[0] == '\\0') {
+            /* handle the numpy.array(b'') case */
```

Comment:
If you're sure we really do want this special case, let's add a clear explanation of what it's for: 
```suggestion
            /* assume that if a user passed `numpy.array(b'')`, which means `numpy.array(b'\0')`, that
            they actually meant `numpy.ndarray((), np.bytes_, b'')`.
            
            Note that as a consequence of this it is impossible to pass a string consisting of a single
            null byte, even though it is possible to pass a string consisting of two null bytes. */
```

Reply:
Let me think about this a little more before the final decision..
It also depends on whether
```
>>> np.array(b'').dtype == np.dtype('S1')
True
```
is considered a numpy bug or a feature?

Reply:
For instance, there is this inconsistency in-between `bytes` and ndarray bytes:
```
>>> np.array(b'\00') == np.array(b'')
array(True)
>>> b'\00' == b''
False
```
that is confusing, IMO.

Reply:
@pearu but here `np.array(b"\00")` is the "suprising" (or broken) behaviour.  You are right it is very hard to create a proper "length 0 string" (for 0-D arrays), but basically, if the user provides an array, it would be correct to ignore 0-bytes on the right until the first non-zero byte is found:
```
>>> np.array(b"\00a\00\00")
array(b'\x00a', dtype='|S4')
```
It is super confusing though, since `bytes(arr)` actually *includes* the zero padding, since it works directly with the array data.  But I think the choice is: either include all trailing `\0` included in the array, or always strip all trailing zeros.

Reply:
I am inclining to remove handling the `numpy.array(b'')` case here because
- f2py should not workaround numpy bugs nor features
- the input to f2py generated code (or any other code) should be treated as it is; the code should not try to guess how the input object was created nor its intention.

Reply:
Removed the special case handling.

Reply:
Removing it sounds good to me.